src/runner.rs

```diff
@@ -3,110 +3,132 @@ use std::ops;
 use std::{u32, usize};
 use std::fmt;
 use std::iter::repeat;
-use std::collections::{HashMap, VecDeque};
-use parity_wasm::elements::{Opcode, BlockType, Local};
+use parity_wasm::elements::Local;
 use {Error, Trap, TrapKind, Signature};
 use module::ModuleRef;
+use memory::MemoryRef;
 use func::{FuncRef, FuncInstance, FuncInstanceInternal};
 use value::{
 	RuntimeValue, FromRuntimeValue, WrapInto, TryTruncateInto, ExtendInto,
 	ArithmeticOps, Integer, Float, LittleEndianConvert, TransmuteInto,
 };
 use host::Externals;
-use common::{DEFAULT_MEMORY_INDEX, DEFAULT_TABLE_INDEX, BlockFrame, BlockFrameType};
-use common::stack::StackWithLimit;
+use common::{DEFAULT_MEMORY_INDEX, DEFAULT_TABLE_INDEX};
 use memory_units::Pages;
 use nan_preserving_float::{F32, F64};
+use isa;
 
 /// Maximum number of entries in value stack.
-pub const DEFAULT_VALUE_STACK_LIMIT: usize = 16384;
-/// Maximum number of entries in frame stack.
-pub const DEFAULT_FRAME_STACK_LIMIT: usize = 16384;
+pub const DEFAULT_VALUE_STACK_LIMIT: usize = (512 * 1024) / ::std::mem::size_of::<RuntimeValue>();
 
-/// Function interpreter.
-pub struct Interpreter<'a, E: Externals + 'a> {
-	externals: &'a mut E,
-}
+// TODO: Make these parameters changeble.
+pub const DEFAULT_CALL_STACK_LIMIT: usize = 16 * 1024;
 
 /// Interpreter action to execute after executing instruction.
 pub enum InstructionOutcome {
 	/// Continue with next instruction.
 	RunNextInstruction,
-	/// Branch to given frame.
-	Branch(usize),
+	/// Branch to an instruction at the given position.
+	Branch(isa::Target),
 	/// Execute function call.
 	ExecuteCall(FuncRef),
-	/// End current frame.
-	End,
 	/// Return from current function block.
-	Return,
+	Return(isa::DropKeep),
```

Comment:
Can the `drop`/`keep` instructions be encoded as extra instructions at the end of the function (so that we don't have to have special-case handling)? That would mean we can make `Branch` and `Return` the same here.

Reply:
What kind of special-case handling do you mean? Also, btw, `drop`/`keep` are not instructions per se, but immediate arguments of these instructions.

Reply:
Special case handling of returning vs branching. Especially in a stack machine you shouldn't need to have a difference.

Reply:
Oh wow, that's an interesting idea, haven't thought about it. It requires some substantial changes though. 

Firstly, all branches currently are static, i.e. they encode branch destination (and drop keep) as immediates. So we need an instruction that pops a return address and branches to it. (It seems that the only instruction that needs such kind of indirect jump, so why don't keep calling it return? : ) )

Secondly, all code currently contained in a separate vectors. So you can't branch from one function and land in code for another. Although keeping all code in a single vec is more efficient and desired. I didn't go with this way since it might be more error-prone and decided to leave it for later.

Also, when I thought about making drop/keep separate instructions, I decided not to, because large part of interpreters overhead comes from a dispatch. So it seems it is profitable to fuse several instructions into one.

Reply:
When this gets merged I'll write a PR to try to convert it to use a flat vector. We can even generate the bytecode for each individual function in parallel and then flatten it at the end.

You've convinced me that this is fine to keep as a special case, though.